### DISCUSSION TITLE
Pyenv with precompiled binaries?

### DISCUSSION THREAD

**Discussion by Erotemic** [pyenv/pyenv#1872] [Q&A] [answered]
Is there any way to use pyenv (or some other tool) to fetch precompiled python binaries for specific versions? I'm trying to replace conda in my workflow, and the compile time of pyenv is currently a limiting factor in how fast I can setup a machine.

Having the option to simply point at a server where the appropriate binaries (instead of the source tarballs) could be downloaded from would be enough.

**Comment by anton-petrov** [comment]
Your proposition is very good and actual for many people. But, for example, python has no compiled packages for linux,
thus, we need additional resources to build and provide binary packages ourselves.

**Reply by Erotemic** [reply]
Does this boil down to needing a machine to build the binaries and then a server to host them on?

Would it be possible/feasible to build the binaries on a CI and then host them on an AWS S3 free tier? Or would this require a more substantial source of funding?

**Reply by native-api** [reply]
> Your proposition is very good and actual for many people. But, for example, python has no compiled packages for linux,
> thus, we need additional resources to build and provide binary packages ourselves.

Unless we have volunteers? https://github.com/pyenv/pyenv/issues/1417

**Reply by native-api** [reply]
The downside of "portable" builds though is that they are statically linked -- so you lose free upgrades for dependencies. For stuff like OpenSSL, that's bad since you'll eventually become vulnerable to TLS attacks and ultimately lose Internet connectivity altogether (it will also have its own certificate store disconnected from the system's one -- though that can be worked around).

**Comment by native-api** [comment]
Travis CI actually precompiles Python versions with Pyenv, then `tar`s the `/versions/<v>` directory and uploads to their hosting: https://github.com/travis-ci/cpython-builder
Then in a build, they just download and unpack it: https://github.com/travis-ci/travis-build/blob/5f10098154313ca99ed72dbf75c7bdf670708f2b/lib/travis/build/script/python.rb#L116-L136
However, they're relying on the fact that all their builders have the same configuration, including preinstalled dependencies (and yes, [they make separate packages for their different environments](https://docs.travis-ci.com/user/languages/python/#python-versions)).

A precompiled binary is going to be distro-specific, so you'll better look for distro-specific resources -- e.g. https://launchpad.net/~deadsnakes/+archive/ubuntu/ppa for Ubuntu.

**Comment by Erotemic** [comment, marked as answer]
Related: 

https://github.com/pyenv/pyenv/issues/2334
https://github.com/orgs/pyenv/discussions/2900#discussion-6193363

**Comment by laggardkernel** [comment]
Use `conda` to download precompiled CPython and install it into the pyenv location.

1. prepend ~/.pyenv/versions folder into conda envs path, where new venv created by conda will be put into.

```bash
export CONDA_ENVS_PATH="${PYENV_ROOT_:-$HOME/.pyenv}/versions:${CONDA_ENVS_PATH_:-${HOME}/.conda/envs}"
```

2. create venv using conda and make the venv name compatible with pyenv.

```bash
conda create -n 3.12.3 python==3.12.3
```

3. check CPython 3.12.3 in pyenv with `pyenv versions`.
